### DISCUSSION TITLE
Raising RunTimeError while using with another async library

### DISCUSSION THREAD

**Discussion by N-i-c-k-007** [fastapi/fastapi#8746] [Questions] [answered]
### First check

* [ x] I added a very descriptive title to this issue.
* [ x] I used the GitHub search to find a similar issue and didn't find it.
* [ x] I searched the FastAPI documentation, with the integrated search.
* [ x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
import coc
import uvicorn
from fastapi import FastAPI

client = coc.login(....)
app = FastAPI()


@app.get("/player")
async def get_player():
    try:
        player = await client.get_player('2pp')
    except coc.NotFound:
        ...
    return player


if __name__ == "__main__":
    uvicorn.run(app)
```

### Description
* Open the browser and call the endpoint `/player`.
* It returns `Internal Server Error`.
* But I expected it to return `player` object.
* Full traceback
```Python
INFO:     127.0.0.1:54122 - "GET /player HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "D:\Projects\Testing\venv\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 369, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "D:\Projects\Testing\venv\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 59, in __call__
    return await self.app(scope, receive, send)
  File "D:\Projects\Testing\venv\lib\site-packages\fastapi\applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\middleware\errors.py", line 181, in __call__
    raise exc from None
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\middleware\errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\exceptions.py", line 82, in __call__
    raise exc from None
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "D:\Projects\Testing\venv\lib\site-packages\starlette\routing.py", line 52, in app
    response = await func(request)
  File "D:\Projects\Testing\venv\lib\site-packages\fastapi\routing.py", line 216, in app
    raw_response = await run_endpoint_function(
  File "D:\Projects\Testing\venv\lib\site-packages\fastapi\routing.py", line 149, in run_endpoint_function
    return await dependant.call(**values)
  File "D:\Projects\Testing\main.py", line 14, in get_player
    player = await client.get_player('2pp')
  File "D:\Projects\Testing\venv\lib\site-packages\coc\client.py", line 1453, in get_player
    data = await self.http.get_player(player_tag)
  File "D:\Projects\Testing\venv\lib\site-packages\coc\http.py", line 255, in request
    async with self.__session.request(method, url, **kwargs) as response:
  File "D:\Projects\Testing\venv\lib\site-packages\aiohttp\client.py", line 1117, in __aenter__
    self._resp = await self._coro
  File "D:\Projects\Testing\venv\lib\site-packages\aiohttp\client.py", line 448, in _request
    with timer:
  File "D:\Projects\Testing\venv\lib\site-packages\aiohttp\helpers.py", line 635, in __enter__
    raise RuntimeError(
RuntimeError: Timeout context manager should be used inside a task
```

### Environment

* OS: `Windows 10`
* FastAPI Version : `0.66.0`
* Python version: `3.9.6`

### Additional context
* If you want to try this code make an account at - https://developer.clashofclans.com/#/
* Replace `coc.login(...)` with `coc.login(email=your email, password=paasword)`, with the above account details.
* Example of creating client instance - [Example](https://github.com/mathsman5133/coc.py/blob/bcc6ff3504107d057c2eb6cd3a329f689bb8721d/examples/discord_bot.py#L21)
* This works fine without the `fastapi`

**Comment by juanbeniteza** [comment]
Try to move the client initialization inside the `async def get_player` method

**Comment by N-i-c-k-007** [comment]
> Try to move the client initialization inside the `async def get_player` method

That doesn't work and it raises `RuntimeWarning: coroutine 'Client.login' was never awaited`. I can't `await` it as it's not coro [Doc Link](https://cocpy.readthedocs.io/en/latest/code_overview/client.html#coc.login)

**Comment by N-i-c-k-007** [comment]
Apparently using `nest_asyncio` solves this. I just did -
```Python
import nest_asyncio
nest_asyncio.apply()
```
and ran it without debugging and it worked. I am not sure if this is something to look into? As it was supposed to work without it

**Comment by Mause** [comment]
Client.login is a coroutine: https://cocpy.readthedocs.io/en/latest/code_overview/client.html#coc.Client.login

**Comment by N-i-c-k-007** [comment]
> Client.login is a coroutine: https://cocpy.readthedocs.io/en/latest/code_overview/client.html#coc.Client.login

That's but we don't use that to get the any instance of the connection, I mean it's simply used to login where as `coc.login` is used for various things [Docs](https://cocpy.readthedocs.io/en/latest/code_overview/client.html#coc.login).
The one at which referred as is being used to login not for that things which can't be done by `coc.login`

**Comment by N-i-c-k-007** [comment]
If i will do something like:
```Python
@app.get("/player")
async def get_player():
    c = await client.login(....)
    try:
        player = await c.get_player('2pp')
    except coc.NotFound:
        ...
    return player
```
then it works fine but -
Doing this doesn't makes any sense as `client` has already returned the connection and if i do this then it's just that i am making a connection over the connection.

**Comment by N-i-c-k-007** [comment]
And this code just solve the issue...
```Python
import coc
import nest_asyncio
from fastapi import FastAPI

nest_asyncio.apply()

client = coc.login(...)
app = FastAPI()


@app.get("/player")
async def get_player():
    try:
        player = await client.get_player('2pp')
    except coc.NotFound:
        return None
    return player.name
```
then `uvicorn main:app --reload` and it works fine. So again, I am not sure if this is something to look into

**Comment by ulan-yisaev** [comment]
Hello,

I'm currently integrating LlamaIndex's OpensearchVectorClient into a FastAPI application and have encountered a `RuntimeError: this event loop is already running` similar to the one described in [Llama Index Issue #12280](https://github.com/run-llama/llama_index/issues/12280). To resolve this, the suggested solution is to use `nest_asyncio.apply()`, which works well in development. However, I'm concerned about potential hidden issues or side effects when deploying this solution in a production environment.

Given the importance of event loop management in async applications and the specific context of FastAPI, I'd appreciate any insights or experiences regarding the use of nest_asyncio in production settings. Are there any known drawbacks, performance impacts, or stability issues that might arise from this approach? Furthermore, are there alternative strategies to manage the event loop in FastAPI that would avoid these issues altogether?

For reference, I'm facing the RuntimeError when attempting to call the OpensearchVectorClient constructor from an async FastAPI endpoint.

**Comment by YuriiMotov** [comment, marked as answer]
If your client needs `async` initialization, do it in `lifespan`:
```python

@asynccontextmanager
async def lifespan(app: FastAPI):
    client = await client.login(....)
    yield {"client": client}
    
    
app = FastAPI(lifespan=lifespan)


async def get_client(request.Request):
    return request.state.client

@app.get("/player")
async def get_player(client = Depends(get_client)):
    try:
        player = await client.get_player('2pp')
    except coc.NotFound:
        ...
    return player

```
